File: android/lib/src/main/java/org/elastos/sdk/elephantwallet/contact/internal/HumanInfo.java
package org.elastos.sdk.elephantwallet.contact.internal;

import com.google.gson.annotations.SerializedName;

import java.util.HashMap;
import java.util.List;

public class HumanInfo {
    public enum Item {
        ChainPubKey(1),
        Did(2),
        ElaAddress(3),
        Nickname(4),
        Avatar(5),
        Gender(6),
        Description(7);

        public static Item valueOf(int id) {
            Item[] values = Item.values();
            for(int idx = 0; idx < values.length; idx++) {
                if(values[idx].id == id) {
                    return values[idx];
                }
            }

            return null;
        }

        int id(){
            return this.id;
        }

        private Item(int id){
            this.id = id;
        }
        private int id;
    }

    public class HumanJson {
        @SerializedName(JsonKey.HumanInfo)
        HumanInfo humanInfo;
    }

    public class CarrierInfo {
        public class DeviceInfo {
            @SerializedName(JsonKey.DeviceId)
            public String devId;
            @SerializedName(JsonKey.DeviceName)
            public String devName;

            @Override
            public String toString() { return Utils.ToString(this); }
        }

        @SerializedName(JsonKey.CarrierAddr)
        public String usrAddr;
        @SerializedName(JsonKey.CarrierId)
        public String usrId;
        @SerializedName(JsonKey.DeviceInfo)
        public DeviceInfo devInfo;
        @SerializedName(JsonKey.UpdateTime)
        public long updateTime;

        @Override
        public String toString() { return Utils.ToString(this); }
    }

    @SerializedName(JsonKey.BoundCarrierArray)
    public List<CarrierInfo> boundCarrierArray;
    @SerializedName(JsonKey.CommonInfoMap)
    private HashMap<Integer, String> commonInfoMap;
    @SerializedName(JsonKey.WalletAddressMap)
    private HashMap<String, String> walletAddressMap;
    @SerializedName(JsonKey.Status)
    public ContactStatus status;
    @SerializedName(JsonKey.HumanCode)
    public String humanCode;
    @SerializedName(JsonKey.UpdateTime)
    public long updateTime;

    public String chainPubKey;
    public String did;
    public String elaAddress;
    public String nickname;
    public String avatar;
    public String gender;
    public String description;

    protected int fromJsonObj(HumanJson info) {
        this.commonInfoMap = null;
        this.walletAddressMap = info.humanInfo.walletAddressMap;
        this.boundCarrierArray = info.humanInfo.boundCarrierArray;
        this.status = info.humanInfo.status;
        this.humanCode = info.humanInfo.humanCode;
        this.updateTime = info.humanInfo.updateTime;

        this.chainPubKey = info.humanInfo.chainPubKey;
        this.did = info.humanInfo.did;
        this.elaAddress = info.humanInfo.elaAddress;
        this.nickname = info.humanInfo.nickname;
        this.avatar = info.humanInfo.avatar;
        this.gender = info.humanInfo.gender;
        this.description = info.humanInfo.description;

        if(info.humanInfo.commonInfoMap != null) {
            this.chainPubKey = info.humanInfo.commonInfoMap.get(Item.ChainPubKey.id());
            this.did = info.humanInfo.commonInfoMap.get(Item.Did.id());
            this.elaAddress = info.humanInfo.commonInfoMap.get(Item.ElaAddress.id());
            this.nickname = info.humanInfo.commonInfoMap.get(Item.Nickname.id());
            this.avatar = info.humanInfo.commonInfoMap.get(Item.Avatar.id());
            this.gender = info.humanInfo.commonInfoMap.get(Item.Gender.id());
            this.description = info.humanInfo.commonInfoMap.get(Item.Description.id());
        }

        return 0;
    }

    protected int toJsonObj(HumanJson json) {
        json.humanInfo = new HumanInfo();

        json.humanInfo.commonInfoMap = null;
        json.humanInfo.walletAddressMap = this.walletAddressMap;
        json.humanInfo.boundCarrierArray = this.boundCarrierArray;
        json.humanInfo.status = this.status;
        json.humanInfo.humanCode = this.humanCode;
        json.humanInfo.updateTime = this.updateTime;

        json.humanInfo.chainPubKey = this.chainPubKey;
        json.humanInfo.did = this.did;
        json.humanInfo.elaAddress = this.elaAddress;
        json.humanInfo.nickname = this.nickname;
        json.humanInfo.avatar = this.avatar;
        json.humanInfo.gender = this.gender;
        json.humanInfo.description = this.description;

        if(this.commonInfoMap != null) {
            json.humanInfo.chainPubKey = this.commonInfoMap.get(Item.ChainPubKey.id());
            json.humanInfo.did = this.commonInfoMap.get(Item.Did.id());
            json.humanInfo.elaAddress = this.commonInfoMap.get(Item.ElaAddress.id());
            json.humanInfo.nickname = this.commonInfoMap.get(Item.Nickname.id());
            json.humanInfo.avatar = this.commonInfoMap.get(Item.Avatar.id());
            json.humanInfo.gender = this.commonInfoMap.get(Item.Gender.id());
            json.humanInfo.description = this.commonInfoMap.get(Item.Description.id());
        }

        return 0;
    }

    protected HumanInfo() {
    }
} // class Factory

File: ios/sdk/swift/internal/HumanInfo.swift

public class HumanInfo: Codable {
  public enum Item: Int {
    case ChainPubKey = 1
    case Did = 2
    case ElaAddress = 3
    case Nickname = 4
    case Avatar = 5
    case Gender = 6
    case Description = 7
  }
  
  public class HumanJson: Codable {
    public var humanInfo: HumanInfo? = nil
    
    enum CodingKeys : String, CodingKey {
      case humanInfo
      
      var rawValue: String {
        get {
          switch self {
            case .humanInfo: return JsonKey.HumanInfo
          }
        }
      }
    }
  }
  
  public class CarrierInfo: Codable {
    public class DeviceInfo: Codable {
      public let devId: String
      public let devName: String
      
      enum CodingKeys : String, CodingKey {
        case devId
        case devName
        
        var rawValue: String {
          get {
            switch self {
              case .devId: return JsonKey.DeviceId
              case .devName: return JsonKey.DeviceName
            }
          }
        }
      }
    }
    
    public let usrAddr: String
    public let usrId: String
    public let devInfo: DeviceInfo
    public let updateTime: Int64
    
    enum CodingKeys : String, CodingKey {
      case usrAddr
      case usrId
      case devInfo
      case updateTime
      
      var rawValue: String {
        get {
          switch self {
            case .usrAddr: return JsonKey.CarrierAddr
            case .usrId: return JsonKey.CarrierId
            case .devInfo: return JsonKey.DeviceInfo
            case .updateTime: return JsonKey.UpdateTime
          }
        }
      }
    }
  }
  
  private(set) var boundCarrierArray: [CarrierInfo]? = nil
  public var commonInfoMap: [Int: String]? = nil
  private var walletAddressMap: [String: String]? = nil
  public private(set) var  status: ContactStatus = .Invalid
  public private(set) var humanCode: String? = nil
  public private(set) var updateTime: Int64 = 0

  enum CodingKeys : String, CodingKey {
    case boundCarrierArray
    case commonInfoMap
    case walletAddressMap
    case status
    case humanCode
    case updateTime
    
    var rawValue: String {
      get {
        switch self {
          case .boundCarrierArray: return JsonKey.BoundCarrierArray
          case .commonInfoMap: return JsonKey.CommonInfoMap
          case .walletAddressMap: return JsonKey.WalletAddressMap
          case .status: return JsonKey.Status
          case .humanCode: return JsonKey.HumanCode
          case .updateTime: return JsonKey.UpdateTime
        }
      }
    }
  }
  
  
  public private(set) var chainPubKey: String? = nil
  public private(set) var did: String? = nil
  public private(set) var elaAddress: String? = nil
  public private(set) var nickname: String? = nil
  public private(set) var avatar: String? = nil
  public private(set) var gender: String? = nil
  public private(set) var description: String? = nil
  
  func fromJsonObj(info: HumanJson) -> Int {
    self.commonInfoMap = nil
    self.walletAddressMap = info.humanInfo?.walletAddressMap
    self.boundCarrierArray = info.humanInfo?.boundCarrierArray
    self.status = info.humanInfo?.status ?? .Invalid
    self.humanCode = info.humanInfo?.humanCode
    self.updateTime = info.humanInfo?.updateTime ?? 0

    self.chainPubKey = info.humanInfo?.chainPubKey
    self.did = info.humanInfo?.did
    self.elaAddress = info.humanInfo?.elaAddress
    self.nickname = info.humanInfo?.nickname
    self.avatar = info.humanInfo?.avatar
    self.gender = info.humanInfo?.gender
    self.description = info.humanInfo?.description

    if info.humanInfo?.commonInfoMap != nil {
      self.chainPubKey = info.humanInfo?.commonInfoMap?[Item.ChainPubKey.rawValue]
      self.did = info.humanInfo?.commonInfoMap?[Item.Did.rawValue]
      self.elaAddress = info.humanInfo?.commonInfoMap?[Item.ElaAddress.rawValue]
      self.nickname = info.humanInfo?.commonInfoMap?[Item.Nickname.rawValue]
      self.avatar = info.humanInfo?.commonInfoMap?[Item.Avatar.rawValue]
      self.gender = info.humanInfo?.commonInfoMap?[Item.Gender.rawValue]
      self.description = info.humanInfo?.commonInfoMap?[Item.Description.rawValue]
    }

    return 0
  }

  func toJsonObj(json: HumanJson) -> Int {
    json.humanInfo = HumanInfo()

    json.humanInfo?.commonInfoMap = nil
    json.humanInfo?.walletAddressMap = self.walletAddressMap
    json.humanInfo?.boundCarrierArray = self.boundCarrierArray
    json.humanInfo?.status = self.status
    json.humanInfo?.humanCode = self.humanCode
    json.humanInfo?.updateTime = self.updateTime

    json.humanInfo?.chainPubKey = self.chainPubKey
    json.humanInfo?.did = self.did
    json.humanInfo?.elaAddress = self.elaAddress
    json.humanInfo?.nickname = self.nickname
    json.humanInfo?.avatar = self.avatar
    json.humanInfo?.gender = self.gender
    json.humanInfo?.description = self.description

    if self.commonInfoMap != nil {
      json.humanInfo?.chainPubKey = self.commonInfoMap?[Item.ChainPubKey.rawValue]
      json.humanInfo?.did = self.commonInfoMap?[Item.Did.rawValue]
      json.humanInfo?.elaAddress = self.commonInfoMap?[Item.ElaAddress.rawValue]
      json.humanInfo?.nickname = self.commonInfoMap?[Item.Nickname.rawValue]
      json.humanInfo?.avatar = self.commonInfoMap?[Item.Avatar.rawValue]
      json.humanInfo?.gender = self.commonInfoMap?[Item.Gender.rawValue]
      json.humanInfo?.description = self.commonInfoMap?[Item.Description.rawValue]
    }

    return 0
  }
  
  init() {
  }
} // class Factory
